Mathlib/CategoryTheory/Limits/Shapes/Pullback/Categorical/Basic.lean

```diff
@@ -0,0 +1,370 @@
+@[reassoc (attr := simp)]
+lemma Hom.w' {x y : F ⊡ G} (f : x ⟶ y) :
+    G.map f.snd ≫ y.iso.inv = x.iso.inv ≫ F.map f.fst := by
+  rw [Iso.comp_inv_eq, Category.assoc, Eq.comm, Iso.inv_comp_eq, f.w]
+
+/-- Extensionnality principle for morphisms in `CategoricalPullback F G`. -/
+@[ext]
+theorem hom_ext {x y : F ⊡ G} {f g : x ⟶ y}
+    (hₗ : f.fst = g.fst) (hᵣ : f.snd = g.snd) : f = g := by
+  apply Hom.ext <;> assumption
+
+section
+
+variable (F G)
+
+/-- `CategoricalPullback.π₁ F G` is the first projection `CategoricalPullback F G ⥤ A`. -/
+@[simps]
+def π₁ : F ⊡ G ⥤ A where
+  obj x := x.fst
+  map f := f.fst
+
+/-- `CategoricalPullback.π₂ F G` is the second projection `CategoricalPullback F G ⥤ C`. -/
+@[simps]
+def π₂ : F ⊡ G ⥤ C where
+  obj x := x.snd
+  map f := f.snd
+
+/-- The canonical categorical commutative square in which `CategoricalPullback F G` sits. -/
+@[simps!]
+instance catCommSq : CatCommSq (π₁ F G) (π₂ F G) F G where
+  iso := NatIso.ofComponents (fun x ↦ x.iso)
+
+variable {F G} in
+/-- Constructor for isomorphisms in `CategoricalPullback F G`. -/
+@[simps!]
+def mkIso {x y : F ⊡ G}
+    (eₗ : x.fst ≅ y.fst) (eᵣ : x.snd ≅ y.snd)
+    (w : F.map eₗ.hom ≫ y.iso.hom = x.iso.hom ≫ G.map eᵣ.hom := by aesop_cat) :
+    x ≅ y where
+  hom := ⟨eₗ.hom, eᵣ.hom, w⟩
+  inv := ⟨eₗ.inv, eᵣ.inv, by simpa using F.map eₗ.inv ≫= w.symm =≫ G.map eᵣ.inv⟩
+
+section
+
+variable {x y : F ⊡ G} (f : x ⟶ y) [IsIso f]
+
+instance : IsIso f.fst :=
+  inferInstanceAs (IsIso ((π₁ _ _).mapIso (asIso f)).hom)
+
+instance : IsIso f.snd :=
+  inferInstanceAs (IsIso ((π₂ _ _).mapIso (asIso f)).hom)
+
+@[simp]
+lemma inv_fst : (inv f).fst = inv f.fst := by
+  symm
+  apply IsIso.inv_eq_of_hom_inv_id
+  simpa [-IsIso.hom_inv_id] using congrArg (fun t ↦ t.fst) (IsIso.hom_inv_id f)
+
+@[simp]
+lemma inv_snd : (inv f).snd = inv f.snd := by
+  symm
+  apply IsIso.inv_eq_of_hom_inv_id
+  simpa [-IsIso.hom_inv_id] using congrArg (fun t ↦ t.snd) (IsIso.hom_inv_id f)
+
+end
+
+lemma isIso_iff {x y : F ⊡ G} (f : x ⟶ y) :
+    IsIso f ↔ (IsIso f.fst ∧ IsIso f.snd) where
+  mp h := ⟨inferInstance, inferInstance⟩
+  mpr | ⟨h₁, h₂⟩ => ⟨⟨inv f.fst, inv f.snd, by aesop_cat⟩, by aesop_cat⟩
+
+end
+
+section
+
+variable (X : Type u₄) [Category.{v₄} X]
+
+variable (F G) in
+/-- The data of a categorical commutative square over a cospan `F, G` with cone point `X` is
+that of a functor `T : X ⥤ A`, a functor `L : X ⥤ C`, and a `CatCommSqOver T L F G`.
+Note that this is *exactly* what an object of
+`((whiskeringRight X A B).obj F) ⊡ ((whiskeringRight X C B).obj G)` is,
+so `CatCommSqOver F G X` is in fact an abbreviation for
+`((whiskeringRight X A B).obj F) ⊡ ((whiskeringRight X C B).obj G)`. -/
+abbrev CatCommSqOver :=
+  (whiskeringRight X A B|>.obj F) ⊡ (whiskeringRight X C B|>.obj G)
+
+namespace CatCommSqOver
+
+/-- Interpret a `CatCommSqOver F G X` as a `CatCommSq`. -/
+@[simps!]
+def asSquare (S : CatCommSqOver F G X) : CatCommSq S.fst S.snd F G where
+  iso := S.iso
```

Comment:
It seems to me the `!` is not necessary, and this should be an instance:
```suggestion
@[simps]
instance asSquare (S : CatCommSqOver F G X) : CatCommSq S.fst S.snd F G where
  iso := S.iso
```

Reply:
Are you sure about making it an instance?
Hopefully, we should not run into too many situations where there is more that one square on the same functors, but I have to admit I’m a bit scared about registering instances for things that might possibly be "non-canonical", and we might create some this way...
I’d prefer treating most the `CatCommSq` as structs, and only register manually those we are sure are canonical (e.g the one in which the categorical pullback sits). In the followup #26412, which also deals with `CatCommSq` mostly as a structure rather than a class (except for a `:= by infer_instance` autoparam), I also intentionally avoided to register any `CatCommSq` instances for this same reason.

Reply:
For `S : CatCommSqOver`, I think it is safe to assume there is a canonical isomorphism!

Reply:
I’ll trust you on this one. Hopefully this will not cause problems when we’ll want to state that flipping the corners of a `CatCommSqOver (Functor.fromPUnit (.unit : SingleObj G)) (Functor.fromPUnit (.unit : SingleObj G)) (Discrete G)` corresponds to taking the inverse of the iso when `[Group G]`...